File: frontend/admin/src/api/modules/seckill.ts
import { type Int64, type Int64Like, type PaginatedList, type PaginationParams } from "@/api/core/types"
import { adminApiClient } from "@/api/core/http"
import { toInt64String } from "@/api/core/int64"

export interface ActivityAdmin {
  activity_id: Int64
  title: string
  description: string
  style_config_json: string
  start_at_unix: number
  end_at_unix: number
  status: number
  created_by: Int64
  updated_by: Int64
  created_at_unix: number
  updated_at_unix: number
  items: ActivityItemAdmin[]
}

export interface ActivityItemAdmin {
  item_id: Int64
  activity_id: Int64
  product_id: Int64
  sku_code: string
  snapshot_name: string
  snapshot_main_image: string
  origin_price_cent: number
  seckill_price_cent: number
  reserved_stock_total: number
  available_stock: number
  sold_stock: number
  user_limit_mode: number
  user_limit_window_sec: number
  user_limit_qty: number
  max_qty_per_order: number
  status: number
  created_at_unix: number
  updated_at_unix: number
}

export interface ActivityOrderView {
  link_id: Int64
  order_id: Int64
  order_no: string
  user_id: Int64
  activity_id: Int64
  activity_item_id: Int64
  quantity: number
  order_status: number
  payment_status: number
  close_reason: string
  last_synced_at_unix: number
  created_at_unix: number
  updated_at_unix: number
}

export interface TrafficBucket {
  bucket_minute_unix: number
  activity_id: Int64
  activity_item_id: Int64
  pv: number
  uv: number
  click: number
  purchase_attempt: number
  purchase_success: number
  purchase_fail: number
  pay_success: number
  order_closed: number
}

export interface ActivityUpsertReq {
  title: string
  description: string
  style_config_json?: string
  start_at_unix: number
  end_at_unix: number
}

export interface UpsertActivityItemReq {
  product_id: Int64Like
  seckill_price_cent: number
  reserved_stock_total: number
  user_limit_mode: number
  user_limit_window_sec: number
  user_limit_qty: number
  max_qty_per_order: number
  status: number
}

export interface ListActivitiesQuery extends PaginationParams {
  keyword?: string
  status?: number
  include_deleted?: boolean
}

export interface GetActivityTrafficQuery {
  activity_item_id?: Int64Like
  from_minute_unix: number
  to_minute_unix: number
}

export interface ListActivityOrdersQuery extends PaginationParams {
  order_status?: number
  payment_status?: number
  user_id?: Int64Like
}

export const adminSeckillApi = {
  createActivity(req: ActivityUpsertReq): Promise<{ activity: ActivityAdmin }> {
    return adminApiClient.post<{ activity: ActivityAdmin }, ActivityUpsertReq>("/seckill/activities", req)
  },

  updateActivity(activityId: Int64Like, req: ActivityUpsertReq): Promise<{ activity: ActivityAdmin }> {
    return adminApiClient.patch<{ activity: ActivityAdmin }, ActivityUpsertReq>(
      `/seckill/activities/${activityId}`,
      req
    )
  },

  deleteActivity(activityId: Int64Like): Promise<{ activity_id: Int64 }> {
    return adminApiClient.delete<{ activity_id: Int64 }>(`/seckill/activities/${activityId}`)
  },

  getActivity(activityId: Int64Like): Promise<{ activity: ActivityAdmin }> {
    return adminApiClient.get<{ activity: ActivityAdmin }>(`/seckill/activities/${activityId}`)
  },

  listActivities(query?: ListActivitiesQuery): Promise<PaginatedList<ActivityAdmin>> {
    return adminApiClient.get<PaginatedList<ActivityAdmin>>("/seckill/activities", {
      params: query,
    })
  },

  createActivityItem(activityId: Int64Like, req: UpsertActivityItemReq): Promise<{ item: ActivityItemAdmin }> {
    return adminApiClient.post<
      { item: ActivityItemAdmin },
      Omit<UpsertActivityItemReq, "product_id"> & { product_id: string }
    >(
      `/seckill/activities/${activityId}/items`,
      {
        ...req,
        product_id: toInt64String(req.product_id, "product_id"),
      }
    )
  },

  updateActivityItem(
    activityId: Int64Like,
    itemId: Int64Like,
    req: UpsertActivityItemReq
  ): Promise<{ item: ActivityItemAdmin }> {
    return adminApiClient.put<
      { item: ActivityItemAdmin },
      Omit<UpsertActivityItemReq, "product_id"> & { product_id: string }
    >(
      `/seckill/activities/${activityId}/items/${itemId}`,
      {
        ...req,
        product_id: toInt64String(req.product_id, "product_id"),
      }
    )
  },

  removeActivityItem(activityId: Int64Like, itemId: Int64Like): Promise<{ activity_id: Int64; item_id: Int64 }> {
    return adminApiClient.delete<{ activity_id: Int64; item_id: Int64 }>(
      `/seckill/activities/${activityId}/items/${itemId}`
    )
  },

  publishActivity(activityId: Int64Like): Promise<{ activity: ActivityAdmin }> {
    return adminApiClient.post<{ activity: ActivityAdmin }>(`/seckill/activities/${activityId}/publish`)
  },

  offlineActivity(activityId: Int64Like): Promise<{ activity: ActivityAdmin }> {
    return adminApiClient.post<{ activity: ActivityAdmin }>(`/seckill/activities/${activityId}/offline`)
  },

  getTraffic(activityId: Int64Like, query: GetActivityTrafficQuery): Promise<{ rows: TrafficBucket[] }> {
    return adminApiClient.get<{ rows: TrafficBucket[] }>(`/seckill/activities/${activityId}/traffic`, {
      params: query,
    })
  },

  listOrders(activityId: Int64Like, query?: ListActivityOrdersQuery): Promise<PaginatedList<ActivityOrderView>> {
    return adminApiClient.get<PaginatedList<ActivityOrderView>>(`/seckill/activities/${activityId}/orders`, {
      params: query,
    })
  },
}

File: apps/seckill/rpc/internal/model/seckill.go
// model 包包含相关应用代码。
package model

import "time"

const (
	// ActivityStatusDraft 表示草稿。
	ActivityStatusDraft int8 = 0
	// ActivityStatusPublished 表示已发布。
	ActivityStatusPublished int8 = 1
	// ActivityStatusOffline 表示已下线。
	ActivityStatusOffline int8 = 2
)

const (
	// ItemStatusDisabled 表示活动商品停用。
	ItemStatusDisabled int8 = 0
	// ItemStatusEnabled 表示活动商品启用。
	ItemStatusEnabled int8 = 1
)

const (
	// UserLimitModeNone 表示不限购。
	UserLimitModeNone int8 = 0
	// UserLimitModeWindowCompleted 表示窗口期按已完成订单限购。
	UserLimitModeWindowCompleted int8 = 1
)

const (
	TrafficEventPV              = "pv"
	TrafficEventClick           = "click"
	TrafficEventPurchaseAttempt = "purchase_attempt"
	TrafficEventPurchaseSuccess = "purchase_success"
	TrafficEventPurchaseFail    = "purchase_fail"
	TrafficEventPaySuccess      = "pay_success"
	TrafficEventOrderClosed     = "order_closed"
)

// Activity 表示秒杀活动聚合。
type Activity struct {
	ID              int64
	Title           string
	Description     string
	StyleConfigJSON string
	StartAt         time.Time
	EndAt           time.Time
	Status          int8
	CreatedBy       int64
	UpdatedBy       int64
	CreatedAt       time.Time
	UpdatedAt       time.Time
	DeletedAt       *time.Time
}

// ActivityItem 表示活动商品配置。
type ActivityItem struct {
	ID                 int64
	ActivityID         int64
	ProductID          int64
	SKUCode            string
	SnapshotName       string
	SnapshotMainImage  string
	OriginPriceCent    int64
	SeckillPriceCent   int64
	ReservedStockTotal int64
	AvailableStock     int64
	SoldStock          int64
	UserLimitMode      int8
	UserLimitWindowSec int64
	UserLimitQty       int64
	MaxQtyPerOrder     int64
	Status             int8
	CreatedAt          time.Time
	UpdatedAt          time.Time
}

// OrderLink 表示活动订单关联记录。
type OrderLink struct {
	ID             int64
	OrderID        int64
	OrderNo        string
	UserID         int64
	ActivityID     int64
	ActivityItemID int64
	Quantity       int64
	OrderStatus    int8
	PaymentStatus  int8
	CloseReason    string
	LastSyncedAt   time.Time
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

// OrderStateSync 表示订单状态回流同步模型。
type OrderStateSync struct {
	OrderID        int64
	OrderNo        string
	UserID         int64
	ActivityID     int64
	ActivityItemID int64
	Quantity       int64
	OrderStatus    int8
	PaymentStatus  int8
	CloseReason    string
	LastSyncedAt   time.Time
}

// TrafficBucket 表示分钟级聚合统计。
type TrafficBucket struct {
	BucketMinute    time.Time
	ActivityID      int64
	ActivityItemID  int64
	PV              int64
	UV              int64
	Click           int64
	PurchaseAttempt int64
	PurchaseSuccess int64
	PurchaseFail    int64
	PaySuccess      int64
	OrderClosed     int64
}

// PurchaseReservation 表示抢购预留结果。
type PurchaseReservation struct {
	ActivityID        int64
	ActivityItemID    int64
	ProductID         int64
	Quantity          int64
	RemainStock       int64
	SeckillPriceCent  int64
	SKUCode           string
	SnapshotName      string
	SnapshotMainImage string
}

// TrafficEvent 表示流量事件。
type TrafficEvent struct {
	ActivityID     int64
	ActivityItemID int64
	EventType      string
	UserID         int64
	ClientID       string
	IdempotencyKey string
	OccurredAt     time.Time
}

// PurchaseTask 表示异步建单任务（Redis 预扣成功后入队）。
type PurchaseTask struct {
	ActivityID     int64
	ActivityItemID int64
	UserID         int64
	Quantity       int64
	IdempotencyKey string
	AccessToken    string
	Item           *ActivityItem
	EnqueuedAt     time.Time
}
